### PR TITLE
Optimize installation

### LINKED ISSUE #310: Optimize installation flow
**What would you like to be added**:

1. Split installation script for different purposes, in a stand-alone cluster(for production) or a [kind](https://kind.sigs.k8s.io/) cluster(for development)
2. No logger using export diff KUBECONFIG to switch multi-clusters in the install script

**Why is this needed**:
1. Likes #303, the newcomers may not have a sense for [kind](https://kind.sigs.k8s.io/) when they wanna make a production installation, so we should make it clear in the installation stage
2. Using a native way to switch multi-clusters for `kubectl`

### PR DESCRIPTION
**Pull request by lfbear**
**What type of PR is this?**

**What this PR does / why we need it**:

1. Two kinds of karmada deployment:
    + deploy to local cluster by kind `local-up-karmada.sh`
    + deploy to remote(stand-alone) cluster `remote-up-karmada.sh`
2. Multi-clusters kebuconfig Integrate in one config file, default in:  /root/.kube/karmda.config. use `kubectl config use-context` to switch.
3. Add some tips in installation step

**Which issue(s) this PR fixes**:

Fixes #310

**Special notes for your reviewer**:

### Some tech details:
- Add '--bind-address' and remove 'host' in probe: keep it can pass when deploying to remote(stand-alone) cluster
- The 2nd argument in 'hack/create-cluster.sh' change to optional, default is current KUBECONFIG
- There is no longer to delete the existed kubeconfig file when run 'hack/create-cluster.sh'
- Save all certificate files to /root/.karmada/ for avoiding loss after reboot
- Add function 'util::append_client_kubeconfig' to append kubeconfig to a existed file
- TODO: modifies sync to karmada-bootstrap.sh & deploy-karmada-agent.sh if it needs

**Does this PR introduce a user-facing change?**:
NONE

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -117,52 +117,91 @@ There are several demonstrations of common cases.
 ```
 
 #### 3. Deploy and run karmada control plane:
+
+Choose a way:
+
+##### 3.1. I have not any cluster
```
Comment (kevin-wangzefeng) [inline README.md:123]: nit: It would be helpful adding a link after here, to quickly jump to `3.2. I have present cluster for installing`

Comment (lfbear) [reply README.md:126]: done

File: hack/util.sh [hunk 2]
```diff
@@ -102,6 +102,19 @@ function util::create_certkey {
 EOF
 }
```
Comment (mrlihanbo) [inline hack/util.sh:104]: add more details about the function would make it easier for people to understand

File: hack/deploy-karmada.sh [hunk 3]
```diff
@@ -129,7 +165,7 @@ kubectl apply -f "${REPO_ROOT}/artifacts/deploy/karmada-scheduler.yaml"
...
 
 # make sure all karmada control plane components are ready
-util::wait_pod_ready ${KARMADA_CONTROLLER_LABEL} "karmada-system"
-util::wait_pod_ready ${KARMADA_SCHEDULER_LABEL} "karmada-system"
-util::wait_pod_ready ${KUBE_CONTROLLER_POD_LABEL} "karmada-system"
-util::wait_pod_ready ${KARMADA_WEBHOOK_LABEL} "karmada-system"
+util::wait_pod_ready "${KARMADA_CONTROLLER_LABEL}" "karmada-system"
+util::wait_pod_ready "${KARMADA_SCHEDULER_LABEL}" "karmada-system"
+util::wait_pod_ready "${KUBE_CONTROLLER_POD_LABEL}" "karmada-system"
+util::wait_pod_ready "${KARMADA_WEBHOOK_LABEL}" "karmada-system"
```
Comment (mrlihanbo) [inline hack/deploy-karmada.sh:171]: new line here

File: hack/util.sh (hunk 2 above)
Comment (lfbear) [reply hack/util.sh:104]: done

**Comment by lfbear** [comment]
@mrlihanbo @kevin-wangzefeng all conversation has resolved, PTAL thanks

**Comment by mrlihanbo** [comment]
Good job! I have test the remote-up-karmada.sh script in my cluster and it works.      
Generally looks good to me!

**Review by mrlihanbo — APPROVED** [review]

File: hack/create-cluster.sh [hunk 4]
```diff
@@ -20,15 +20,30 @@ if [[ -z "${CLUSTER_NAME}" ]]; then
...
+source "${REPO_ROOT}"/hack/util.sh
+util::cmd_must_exist "kind"
+
 if [ -f "${KUBECONFIG}" ];then
-  echo "Removing old kubeconfig file."
-  rm -f ${KUBECONFIG}
+  echo "kubeconfig file is existed, new config will append to it."
+  if kubectl config get-contexts "${CLUSTER_NAME}" --kubeconfig="${KUBECONFIG}"> /dev/null 2>&1;
+  then
+    echo "ERROR: failed to create new cluster for context '${CLUSTER_NAME}' existed in ${KUBECONFIG}. please remove it (use 'kubectl config delete-context') if your want to recover it."
```
Comment (RainbowMango) [inline hack/create-cluster.sh:44]: > if your want to recover it.

Just a question, what do you mean `recover`?

File: hack/deploy-karmada.sh [hunk 5]
```diff
@@ -3,59 +3,84 @@
 set -o errexit
 set -o nounset
 
+# This script deploy karmada control plane to any cluster you want.	REPO_ROOT=$(dirname "${BASH_SOURCE[0]}")/..
+# This script intended used in following scenarios:	source ${REPO_ROOT}/hack/util.sh
```
Comment (RainbowMango) [inline hack/deploy-karmada.sh:7]: Do you mean this script depends on utils in `hack/util.sh`?

**Comment by RainbowMango** [comment]
Awesome~

Filed [an issue](https://github.com/karmada-io/karmada/issues/372) to refactor `karmada-bootstrap.sh` after this PR get merged.
@mrlihanbo

File: hack/create-cluster.sh (hunk 4 above)
Comment (lfbear) [reply hack/create-cluster.sh:44]: sorry, typo. it should be 'cover'

File: hack/deploy-karmada.sh (hunk 5 above)
Comment (lfbear) [reply hack/deploy-karmada.sh:7]: yep

**Comment by lfbear** [comment]
@RainbowMango I have resolved the questions, PTAL thank you

**Comment by RainbowMango** [comment]
leave approve to @kevin-wangzefeng

File: hack/remote-up-karmada.sh [hunk 6]
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env bash
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+function usage() {
+  echo "This script will deploy karmada control plane to a given cluster."
+  echo "Usage: hack/deploy-karmada.sh <KUBECONFIG> <CONTEXT_NAME>"
+  echo "Example: hack/deploy-karmada.sh ~/.kube/config karmada-host"
```
[the remark this answers is not in the document]
Comment (lfbear) [reply hack/remote-up-karmada.sh]: Thank you for pointing out the stupid mistake

[the remark this answers is not in the document]
Comment (RainbowMango) [reply hack/remote-up-karmada.sh]: Typo always happens, no worry.

**Comment by RainbowMango** [comment]
@mrlihanbo Do you have any other comments? Since v0.6 is coming soon, I really want this in the new version.

**Comment by mrlihanbo** [comment]
@RainbowMango No others, thanks.

**Comment by lfbear** [comment]
Rebase multi-commits to a single one

**Review by kevin-wangzefeng — APPROVED** [review]
LGTM, thanks.

**Pull request merged**